src/main/java/org/opensearch/geospatial/ip2geo/jobscheduler/DatasourceRunner.java

```diff
@@ -0,0 +1,286 @@
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.geospatial.ip2geo.jobscheduler;
+
+import java.net.URL;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+
+import lombok.extern.log4j.Log4j2;
+
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.support.IndicesOptions;
+import org.opensearch.client.Client;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.geospatial.ip2geo.common.DatasourceHelper;
+import org.opensearch.geospatial.ip2geo.common.DatasourceManifest;
+import org.opensearch.geospatial.ip2geo.common.DatasourceState;
+import org.opensearch.geospatial.ip2geo.common.GeoIpDataHelper;
+import org.opensearch.geospatial.ip2geo.common.Ip2GeoSettings;
+import org.opensearch.jobscheduler.spi.JobExecutionContext;
+import org.opensearch.jobscheduler.spi.ScheduledJobParameter;
+import org.opensearch.jobscheduler.spi.ScheduledJobRunner;
+import org.opensearch.jobscheduler.spi.utils.LockService;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * Datasource update task
+ *
+ * This is a background task which is responsible for updating Ip2Geo datasource
+ */
+@Log4j2
+public class DatasourceRunner implements ScheduledJobRunner {
+
+    private static DatasourceRunner INSTANCE;
+
+    /**
+     * Return a singleton job runner instance
+     * @return job runner
+     */
+    public static DatasourceRunner getJobRunnerInstance() {
+        if (INSTANCE != null) {
+            return INSTANCE;
+        }
+        synchronized (DatasourceRunner.class) {
+            if (INSTANCE != null) {
+                return INSTANCE;
+            }
+            INSTANCE = new DatasourceRunner();
+            return INSTANCE;
+        }
+    }
+
+    private ClusterService clusterService;
+    private ThreadPool threadPool;
+    private Client client;
+    private TimeValue timeout;
+    private int indexingBulkSize;
+
+    private DatasourceRunner() {
+        // Singleton class, use getJobRunner method instead of constructor
+    }
+
+    /**
+     * Set cluster service
+     * @param clusterService the cluster service
+     */
+    public void setClusterService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+
+        timeout = Ip2GeoSettings.TIMEOUT_IN_SECONDS.get(clusterService.getSettings());
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(Ip2GeoSettings.TIMEOUT_IN_SECONDS, it -> timeout = it);
+        indexingBulkSize = Ip2GeoSettings.INDEXING_BULK_SIZE.get(clusterService.getSettings());
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(Ip2GeoSettings.INDEXING_BULK_SIZE, it -> indexingBulkSize = it);
```

Comment:
The function name is setClusterService and after setting the cluster service we are updating the setting consumers and other things. We need to move this out and call that new function separately. The function is doing 2 things at once. It breaks the single responsibility principle.

Reply:
Moved them to `initialize()` method